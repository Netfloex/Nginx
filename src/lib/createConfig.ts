import ConfigParser, { NginxLocation } from "@webantic/nginx-config-parser";
import { join } from "path";

import { baseConf } from "@utils/baseConf";
import { createAuthFile } from "@utils/createAuthFile";
import { createHash } from "@utils/createHash";
import { downloadCSSToFile } from "@utils/downloadCSSToFile";
import { downloadJSToFile } from "@utils/downloadJSToFile";
import { isObjectEmpty } from "@utils/isObjectEmpty";
import settings from "@utils/settings";
import { sslFilesFor } from "@utils/sslFilesFor";

import { SimpleServer } from "@models/ParsedConfig";
import { Locations, Server } from "@models/config";

const parser = new ConfigParser();

const usesCustom = (options: Server): boolean =>
	!!options.custom_css?.length || !!options.custom_js?.length;

const createLocation = async (
	location: Server,
	defaultUsername: string
): Promise<NginxLocation> => {
	const block: NginxLocation = {};
	const promises: Promise<void>[] = [];

	// Proxy Pass
	if (location.proxy_pass) {
		if (settings.dontExitNoUpstream) {
			block.set ??= [];
			block.set.push(`$upstream "${location.proxy_pass}"`);
			block.proxy_pass = "$upstream";
		} else {
			block.proxy_pass = location.proxy_pass;
		}

		block.include ??= [];
		block.include.push(join(settings.nginxIncludePath, "proxy_pass.conf"));
	}

	// Return
	if (location.return) block.return = location.return;

	if (location.redirect) block.return = `301 ${location.redirect}`;

	if (location.rewrite) block.rewrite = location.rewrite;

	if (location.html) {
		block.return = `200 "${location.html}"`;
		location.headers ??= {};
		location.headers["Content-Type"] = "text/html";
	}

	if (location.static) block.root = location.static;

	if (location.websocket) {
		// Websocket
		block.proxy_set_header ??= [];
		block.proxy_set_header.push(
			"Upgrade $http_upgrade",
			"Connection $http_connection"
		);

		block.proxy_http_version = 1.1;
	}

	if (usesCustom(location)) {
		// Custom Files
		block.sub_filter = [];
		block.proxy_set_header ??= [];
		block.sub_filter_once = "on";
		block.proxy_set_header.push('Accept-Encoding ""');

		// Custom CSS
		if (location.custom_css?.length) {
			const fileNames = location.custom_css.map((g) => createHash(g));

			block.sub_filter.push(
				`'</head>' '${fileNames
					.map(
						(hash) =>
							`<link rel="stylesheet" type="text/css" href="/custom_assets/css/${hash}.css">`
					)
					.join("")}</head>'`
			);

			if (!settings.dontDownloadCustomFiles)
				promises.push(downloadCSSToFile(location.custom_css));
		}

		// Custom JS
		if (location.custom_js?.length) {
			const fileNames = location.custom_js.map((g) => createHash(g));

			block.sub_filter.push(
				`'</body>' '${fileNames
					.map(
						(hash) =>
							`<script src="/custom_assets/js/${hash}.js"></script>`
					)
					.join("")}</body>'`
			);

			if (!settings.dontDownloadCustomFiles)
				promises.push(downloadJSToFile(location.custom_js));
		}
	}

	if (location.noindex) {
		location.headers ??= {};
		location.headers["X-Robots-Tag"] =
			"'noindex, nofollow, nosnippet, noarchive'";
	}

	// Headers
	const headerEntries = Object.entries(location.headers ?? {});

	if (headerEntries.length) {
		block.add_header = headerEntries.map((header) => header.join(" "));
	}

	// Auth

	if (location.auth) {
		promises.push(
			(async (): Promise<void> => {
				const { filepath, hash } = await createAuthFile(
					location.auth!,
					defaultUsername
				);
				block.auth_basic = hash;
				block.auth_basic_user_file = filepath;
			})()
		);
	}

	// Raw

	if (location.raw) {
		Object.entries(location.raw).forEach(([key, value]) => {
			block[key] ??= [];
			block[key].push(...value);
		});
	}

	// Include

	if (location.include) {
		block.include ??= [];
		block.include.push(...location.include);
	}

	await Promise.all(promises);

	return block;
};

const createConfig = async (
	server: Omit<SimpleServer, "filename">,
	defaultUsername = "admin"
): Promise<string> => {
	const { server: jsonServer } = await baseConf();

	// Server Name
	jsonServer.server_name = server.server_name;

	jsonServer.listen = [
		`${server.port} ${server.ssl ? `ssl http2` : ""}`,
		`[::]:${server.port}`
	];

	// SSL Certificate files
	if (!server.disable_cert) {
		const files = sslFilesFor(server);
		Object.assign(jsonServer, files);
	}

	if (settings.dontExitNoUpstream) {
		jsonServer.resolver = settings.dnsResolver;
	}

	const locationSlash = await createLocation(
		{
			...server,
			location: "/"
		} as Locations[0],
		defaultUsername
	);
	if (!isObjectEmpty(locationSlash)) jsonServer["location /"] = locationSlash;

	// Custom Locations
	if (server.locations?.length) {
		await Promise.all(
			server.locations.map(async (location) => {
				const namedLocation = await createLocation(
					location,
					defaultUsername
				);
				if (!isObjectEmpty(namedLocation))
					jsonServer[`location ${location.location}`] = namedLocation;
			})
		);
	}

	if (usesCustom(server)) {
		jsonServer["location /custom_assets"] = {};
		jsonServer["location /custom_assets"].alias = settings.customFilesPath;
	}

	const config = parser.toConf({
		server: jsonServer
	});

	return config;
};

export default createConfig;
