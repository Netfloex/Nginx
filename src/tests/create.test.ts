import { replaceCurrentDir } from "./test-utils";

import createConfig from "@lib/createConfig";
import settings from "@utils/settings";

import fullValidServer from "@configs/full-valid-server.json";

describe("Create config", () => {
	test("Base Config", async () => {
		expect(
			replaceCurrentDir(
				await createConfig({
					server_name: "example.com",
					port: 443,
					ssl: true
				})
			)
		).toMatchSnapshot();
	});

	test("Full Server", async () => {
		settings.dontDownloadCustomFiles = true;
		expect(
			replaceCurrentDir(await createConfig(fullValidServer))
		).toMatchSnapshot();
	});
});
