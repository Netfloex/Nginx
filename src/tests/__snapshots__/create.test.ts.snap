// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create config Base Config 1`] = `
"server {
    listen                     443 ssl http2;
    listen                     [::]:443;
    server_name                example.com;
    ssl_certificate            /etc/letsencrypt/live/example.com/fullchain.pem;
    ssl_certificate_key        /etc/letsencrypt/live/example.com/privkey.pem;
    ssl_trusted_certificate    /etc/letsencrypt/live/example.com/chain.pem;
    ssl_dhparam                /etc/letsencrypt/dhparams/dhparam.pem;
}

"
`;

exports[`Create config Full Server 1`] = `
"server {
    listen                     443 ssl http2;
    listen                     [::]:443;
    server_name                example.com;
    ssl_certificate            /etc/letsencrypt/live/certbot/fullchain.pem;
    ssl_certificate_key        /etc/letsencrypt/live/certbot/privkey.pem;
    ssl_trusted_certificate    /etc/letsencrypt/live/certbot/chain.pem;
    ssl_dhparam                /etc/letsencrypt/dhparams/dhparam.pem;
    location / {
        proxy_pass              http://example;
        include                 /current/src/nginx/builtin/base/proxy_pass.conf;
        return                  200 \\"html\\";
        rewrite                 ^ https://rewrite;
        proxy_set_header        Upgrade $http_upgrade;
        proxy_set_header        Connection $http_connection;
        proxy_set_header        Accept-Encoding \\"\\";
        proxy_http_version      1.1;
        sub_filter              '</head>' '<link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/custom_assets/css/a453d6427958b58add4ed834d9c6520d.css\\"><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/custom_assets/css/e504c8433e0e5044eda93a31b93abd11.css\\"></head>';
        sub_filter              '</body>' '<script src=\\"/custom_assets/js/78f5655ab86382e62279a4dcea0a6bab.js\\"></script><script src=\\"/custom_assets/js/e504c8433e0e5044eda93a31b93abd11.js\\"></script></body>';
        sub_filter_once         on;
        add_header              Content-Type text/html;
        add_header              Second Value;
        add_header              X-Robots-Tag 'noindex, nofollow, nosnippet, noarchive';
        custom                  option;
        auth_basic              2d77215cd91113df4d3442ee904fd740;
        auth_basic_user_file    /current/data/auth/2d77215cd91113df4d3442ee904fd740;
    }

    location /custom_assets {
        alias    /current/data/custom;
    }

}

"
`;
